backends/sqlite/sqlite_backend.go

```diff
@@ -0,0 +1,768 @@
+		return
+	}
+
+	// check if the job is being retried and increment retry count accordingly
+	if job.Status != internal.JobStatusNew {
+		job.Retries++
+	}
+
+	h, ok := s.handlers[job.Queue]
+	if !ok {
+		s.logger.Error("received a job for which no handler is configured",
+			slog.String("queue", job.Queue),
+			slog.Int64("job_id", job.ID))
+		return handler.ErrNoHandlerForQueue
+	}
+
+	ctx = withJobContext(ctx, job)
+
+	if job.Deadline != nil && job.Deadline.Before(time.Now().UTC()) {
+		err = jobs.ErrJobExceededDeadline
+		s.logger.Debug("job deadline is in the past, skipping", slog.String("queue", job.Queue), slog.Int64("job_id", job.ID))
+
+		err = s.updateJobToFailedByDeadline(ctx, h, err, job)
+		if err != nil {
+			s.logger.Error("unable to update job status", "error", err, "job_id", job.ID)
+		}
+		return
+	}
+
+	err = s.updateJobToInProgress(ctx, h, job)
+	if err != nil {
+		return
+	}
+
+	// execute the queue handler of this job
+	jobErr := handler.Exec(ctx, h)
+
+	err = s.updateJobAfterExecution(ctx, h, jobErr, job)
+	return
+}
+
+func (s *SqliteBackend) updateJobAfterExecution(ctx context.Context, h handler.Handler, jobErr error, job *jobs.Job) (err error) {
+	var tx *sql.Tx
+
+	s.dbMutex.Lock()
+	defer s.dbMutex.Unlock()
+
+	tx, err = s.db.Begin()
+	if err != nil {
+		return
+	}
+	defer func() { _ = tx.Rollback() }() // rollback has no effect if the transaction has been committed
+
+	ctx = context.WithValue(ctx, txCtxVarKey, tx)
+
+	err = s.updateJob(ctx, jobErr, "")
+	if err != nil {
+		if errors.Is(err, context.Canceled) {
+			return
+		}
+		err = fmt.Errorf("error updating job status: %w", err)
+		return err
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		errMsg := "unable to commit job transaction. retrying this job may duplicate work:"
+		s.logger.Error(errMsg, slog.String("queue", h.Queue), slog.Any("error", err), slog.Int64("job_id", job.ID))
+		return fmt.Errorf("%s %w", errMsg, err)
+	}
+
+	return
+}
+
+func (s *SqliteBackend) updateJobToFailedByDeadline(ctx context.Context, h handler.Handler, jobErr error, job *jobs.Job) (err error) {
+	var tx *sql.Tx
+
+	s.dbMutex.Lock()
+	defer s.dbMutex.Unlock()
+
+	tx, err = s.db.Begin()
+	if err != nil {
+		return
+	}
+
+	ctx = context.WithValue(ctx, txCtxVarKey, tx)
+
+	err = s.updateJob(ctx, jobErr, internal.JobStatusInProgress)
+	if err != nil {
+		if errors.Is(err, context.Canceled) {
+			return
+		}
+		err = fmt.Errorf("error updating job status: %w", err)
+		return err
+	}
+	err = tx.Commit()
+	if err != nil {
+		errMsg := "unable to commit job transaction. retrying this job may duplicate work:"
+		s.logger.Error(errMsg, slog.String("queue", h.Queue), slog.Any("error", err), slog.Int64("job_id", job.ID))
+		_ = tx.Rollback()
+		return fmt.Errorf("%s %w", errMsg, err)
+	}
+	return
+}
+
+func (s *SqliteBackend) updateJobToInProgress(ctx context.Context, h handler.Handler, job *jobs.Job) (err error) {
```

Comment:
I think we should consider having this function receive a `*sql.Tx` rather than it create one. That way, job handling can use a single transaction for all db operations.

In the PG backend, we do this by adding a `tx` to the `ctx`, e.g. from `handleJob`:

```
ctx = context.WithValue(ctx, txCtxVarKey, tx)
```

Reply:
In any case we'll need atleast 2 transactions right? One for `in progress/failed` and one for `processed/failed` ? Since `in progress` needs to be committed

Reply:
How important is `in progress` to you? Is there a specific use case you have in mind? 

I intentionally avoided it for postgres because it's an additional database round trip, and the best use case I had for it was to be able to show in-progress jobs in a UI (which doesn't exist). As you can see, leaving in progress out also reduced complexity because the "state machine" for determining its correct value also needs to be considered. I'd be inclined to remove the `in progress` requirement unless it's serving a specific goal of yours.

Reply:
I had the very same use case for being able to show in-progress jobs in a UI. But maybe you're right, it adds more complexity. We can perhaps remove the requirement for now as it would also be consistent with other backends.